tests/unit/test_util.py

```diff
@@ -78,69 +83,39 @@ def test_pairwise_difference(self) -> None:
         x_array = generator.random((num_points_x, dimension))
         y_array = generator.random((num_points_y, dimension))
         expected_output = np.array([[x - y for y in y_array] for x in x_array])
-        output = coreax.util.pairwise_difference(x_array, y_array)
-        self.assertAlmostEqual(
-            float(jnp.linalg.norm(output - expected_output)), 0.0, places=3
-        )
+        output = pairwise(difference)(x_array, y_array)
+        assert jnp.linalg.norm(output - expected_output) == pytest.approx(0.0, abs=1e-3)
 
     def test_apply_negative_precision_threshold_invalid(self) -> None:
         """
-        Test the function apply_negative_precision_threshold with an invalid threshold.
+        Test apply_negative_precision_threshold for an invalid threshold.
 
         A negative precision threshold is given, which should be rejected by the
         function.
         """
-        self.assertRaises(
-            ValueError,
-            coreax.util.apply_negative_precision_threshold,
-            x=0.1,
-            precision_threshold=-1e-8,
-        )
-
-    def test_apply_negative_precision_threshold_valid_no_change(self) -> None:
-        """
-        Test the function apply_negative_precision_threshold with no change needed.
-
-        This test questions if the value -0.01 is sufficiently close to 0 to set it to
-        0, however the precision threshold is sufficiently small to consider this a
-        distinct value and not apply a cap.
-        """
-        func_out = coreax.util.apply_negative_precision_threshold(
-            x=-0.01, precision_threshold=0.001
-        )
-        self.assertEqual(func_out, -0.01)
-
-    def test_apply_negative_precision_threshold_valid_with_change(self) -> None:
-        """
-        Test the function apply_negative_precision_threshold with a change needed.
-
-        This test questions if the value -0.0001 is sufficiently close to 0 to set it to
-        0. In this instance, the precision threshold is sufficiently large to consider
-        -0.0001 close enough to 0 to apply a cap.
-        """
-        func_out = coreax.util.apply_negative_precision_threshold(
-            x=-0.0001, precision_threshold=0.001
-        )
-        self.assertEqual(func_out, 0.0)
-
-    def test_apply_negative_precision_threshold_valid_positive_input(self) -> None:
+        with pytest.raises(ValueError):
+            apply_negative_precision_threshold(x=0.1, precision_threshold=-1e-8)
+
+    @pytest.mark.parametrize(
+        "value, threshold, expected",
+        [
+            (-0.01, 0.001, -0.01),
+            (-0.0001, 0.001, 0.0),
+            (0.01, 0.001, 0.01),
+            (0.000001, 0.001, 0.000001),
+        ],
+        ids=["no_change", "with_change", "positive_input_1", "positive_input_2"],
+    )
+    def test_apply_negative_precision_threshold(
```

Comment:
Part of the pull description is to utilise pytest in the utils tests. We could separate this out into another pull, but as it was only a small change to the same tests `test_utils` that needed updating anyway, I thought it would be easier to include here.